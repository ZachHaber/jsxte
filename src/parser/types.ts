import type { ElementStruct } from "../jsx/jsx.types";
import type { EventCallback } from "./base-html-parser/events.types";

export type HTMLElementStruct = {
  tag: string;
  key?: any;
  children: ElementStruct[];
  attributes: [string, string | undefined][];
  events: [string, EventCallback][];
  text: string | undefined;
};

export type Rewrap<T extends object> = T extends infer OBJ
  ? {
      [K in keyof OBJ]: OBJ[K] extends infer O ? O : never;
    }
  : never;

export type HTMLPropsFor<Attributes, Events> = {
  [K in keyof Attributes]?: Attributes[K];
} & {
  [K in keyof Events]?: (e: Events[K]) => void;
};
