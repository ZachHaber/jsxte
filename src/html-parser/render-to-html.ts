import { jsxElemToHtmlAsync, jsxElemToHtmlSync } from "./jsx-elem-to-html";

/**
 * Renders the provided JSX component to pure html. This function
 * is synchronous and will not render components that are asynchronous.
 */
export const renderToHtml = (component: JSX.Element) => {
  return jsxElemToHtmlSync(component);
};

/**
 * Renders the provided JSX component to pure html. This function
 * is asynchronous, it allows to use asynchronous components
 * withing the provided component structure.
 */
export const renderToHtmlAsync = async (
  component: JSX.Element | Promise<JSX.Element>
) => {
  return await jsxElemToHtmlAsync(await component);
};
