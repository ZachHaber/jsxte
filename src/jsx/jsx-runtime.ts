import type { ElemOrList } from "./jsx.types";

type CreateElementProps = {
  [k: string]: any;
  children?: ElemOrList<JSX.Element | string> | undefined;
};

export const createElement = (
  tag: string,
  props?: CreateElementProps,
  ...children: Array<ElemOrList<JSX.Element | string>>
): JSX.Element => {
  props ??= {};

  if (children) {
    props.children = [
      ...(props.children
        ? Array.isArray(props.children)
          ? props.children
          : [props.children]
        : []),
      ...children.flat(),
    ];
  }

  if (props?.children) {
    if (typeof props.children === "string") {
      props.children = { type: "textNode", text: props.children };
    } else if (Array.isArray(props.children)) {
      props.children = props.children.map((child): JSX.Element => {
        if (typeof child === "string") {
          return { type: "textNode", text: child };
        }
        return child;
      });
    }
  }

  return {
    type: "tag",
    tag,
    props: props as JSX.ElementProps,
  };
};

export const jsx = createElement;
export const jsxs = jsx;
export const _jsx = jsx;
export const _jsxs = jsx;

export const Fragment = "";
export const _Fragment = Fragment;
