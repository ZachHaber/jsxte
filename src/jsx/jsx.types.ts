import type { BaseHTMLProps } from "../parser/base-html-parser/base-html-jsx-props";
import type { AnchorProps } from "../parser/html-tag-parsers/a/a-jsx-props";
import type { ButtonProps } from "../parser/html-tag-parsers/button/button-jsx-props";
import type { InputProps } from "../parser/html-tag-parsers/input/input-jsx-props";

export type ElemOrList<T> = T | T[];

export type ElementStructChildren = ElemOrList<ElementStruct | string>;

export type ElementStructProps = {
  key?: any;
  children?: ElementStructChildren;
  [k: string]: any;
};

export type ElementStruct = {
  tag: string | ((props: ElementStructProps) => ElementStruct);
  props: ElementStructProps;
};

declare global {
  namespace JSX {
    type Element = ElementStruct;
    type IntrinsicElements = {
      a: BaseHTMLProps<AnchorProps>;
      abbr: BaseHTMLProps;
      address: BaseHTMLProps;
      area: BaseHTMLProps;
      article: BaseHTMLProps;
      aside: BaseHTMLProps;
      audio: BaseHTMLProps;
      b: BaseHTMLProps;
      base: BaseHTMLProps;
      bdi: BaseHTMLProps;
      bdo: BaseHTMLProps;
      big: BaseHTMLProps;
      blockquote: BaseHTMLProps;
      body: BaseHTMLProps;
      br: BaseHTMLProps;
      button: BaseHTMLProps<ButtonProps>;
      canvas: BaseHTMLProps;
      caption: BaseHTMLProps;
      cite: BaseHTMLProps;
      code: BaseHTMLProps;
      col: BaseHTMLProps;
      colgroup: BaseHTMLProps;
      data: BaseHTMLProps;
      datalist: BaseHTMLProps;
      dd: BaseHTMLProps;
      del: BaseHTMLProps;
      details: BaseHTMLProps;
      dfn: BaseHTMLProps;
      dialog: BaseHTMLProps;
      div: BaseHTMLProps;
      dl: BaseHTMLProps;
      dt: BaseHTMLProps;
      em: BaseHTMLProps;
      embed: BaseHTMLProps;
      fieldset: BaseHTMLProps;
      figcaption: BaseHTMLProps;
      figure: BaseHTMLProps;
      footer: BaseHTMLProps;
      form: BaseHTMLProps;
      h1: BaseHTMLProps;
      h2: BaseHTMLProps;
      h3: BaseHTMLProps;
      h4: BaseHTMLProps;
      h5: BaseHTMLProps;
      h6: BaseHTMLProps;
      head: BaseHTMLProps;
      header: BaseHTMLProps;
      hgroup: BaseHTMLProps;
      hr: BaseHTMLProps;
      html: BaseHTMLProps;
      i: BaseHTMLProps;
      iframe: BaseHTMLProps;
      img: BaseHTMLProps;
      input: BaseHTMLProps<InputProps>;
      ins: BaseHTMLProps;
      kbd: BaseHTMLProps;
      keygen: BaseHTMLProps;
      label: BaseHTMLProps;
      legend: BaseHTMLProps;
      li: BaseHTMLProps;
      link: BaseHTMLProps;
      main: BaseHTMLProps;
      map: BaseHTMLProps;
      mark: BaseHTMLProps;
      menu: BaseHTMLProps;
      menuitem: BaseHTMLProps;
      meta: BaseHTMLProps;
      meter: BaseHTMLProps;
      nav: BaseHTMLProps;
      noindex: BaseHTMLProps;
      noscript: BaseHTMLProps;
      object: BaseHTMLProps;
      ol: BaseHTMLProps;
      optgroup: BaseHTMLProps;
      option: BaseHTMLProps;
      output: BaseHTMLProps;
      p: BaseHTMLProps;
      param: BaseHTMLProps;
      picture: BaseHTMLProps;
      pre: BaseHTMLProps;
      progress: BaseHTMLProps;
      q: BaseHTMLProps;
      rp: BaseHTMLProps;
      rt: BaseHTMLProps;
      ruby: BaseHTMLProps;
      s: BaseHTMLProps;
      samp: BaseHTMLProps;
      slot: BaseHTMLProps;
      script: BaseHTMLProps;
      section: BaseHTMLProps;
      select: BaseHTMLProps;
      small: BaseHTMLProps;
      source: BaseHTMLProps;
      span: BaseHTMLProps;
      strong: BaseHTMLProps;
      style: BaseHTMLProps;
      sub: BaseHTMLProps;
      summary: BaseHTMLProps;
      sup: BaseHTMLProps;
      table: BaseHTMLProps;
      template: BaseHTMLProps;
      tbody: BaseHTMLProps;
      td: BaseHTMLProps;
      textarea: BaseHTMLProps;
      tfoot: BaseHTMLProps;
      th: BaseHTMLProps;
      thead: BaseHTMLProps;
      time: BaseHTMLProps;
      title: BaseHTMLProps;
      tr: BaseHTMLProps;
      track: BaseHTMLProps;
      u: BaseHTMLProps;
      ul: BaseHTMLProps;
      var: BaseHTMLProps;
      video: BaseHTMLProps;
      wbr: BaseHTMLProps;
      webview: BaseHTMLProps;

      // SVG
      svg: BaseHTMLProps;
    };
  }
}
