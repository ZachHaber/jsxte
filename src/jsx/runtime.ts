export const createElement = (
  tag: string,
  props: Record<string, any>, // TODO: typings
  key?: any
): JSX.Element => {
  props.key = key;
  return {
    tag,
    props,
  };
};

export const jsx = createElement;
export const jsxs = jsx;
export const _jsx = jsx;
export const _jsxs = jsx;

export const Fragment = "";
export const _Fragment = Fragment;
