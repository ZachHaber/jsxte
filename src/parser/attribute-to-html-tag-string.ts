import type { RendererHTMLAttributes } from "./types";

export const attributeToHtmlTagString = ([key, value]: [
  string,
  string | undefined
]): string => {
  if (value === undefined) {
    return `${key}`;
  }
  return `${key}="${value.replace('"', '\\"')}"`;
};

export const mapAttributesToHtmlTagString = (
  attributes: RendererHTMLAttributes
): string[] => {
  return attributes.map(attributeToHtmlTagString);
};
