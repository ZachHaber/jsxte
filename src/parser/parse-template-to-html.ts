import { pad } from "../utilities/pad";
import { mapAttributesToHtmlTagString } from "./attribute-to-html-tag-string";
import { getHTMLStruct } from "./get-html-struct";

export const jsxElemToHTML = (element: JSX.Element, indent = 0): string => {
  if (element.type === "textNode") {
    const indentPadding = pad(indent);
    return indentPadding + element.text;
  }

  if (typeof element.tag !== "string") {
    if (typeof element.tag !== "function") {
      console.log(element);
    }
    const subElem = element.tag(element.props);
    return jsxElemToHTML(subElem, indent);
  } else {
    const htmlStruct = getHTMLStruct(element);

    if (htmlStruct.tag === "") {
      const results: string[] = [];
      for (const child of htmlStruct.children) {
        results.push(jsxElemToHTML(child, indent + 2));
      }
      return results.join("\n");
    } else {
      const indentPadding = pad(indent);

      const startTag =
        [
          `${indentPadding}<${htmlStruct.tag}`,
          ...mapAttributesToHtmlTagString(htmlStruct.attributes),
        ].join(" ") + ">";
      const endTag = `${indentPadding}</${htmlStruct.tag}>`;
      const children: string[] = [];

      for (const child of htmlStruct.children) {
        children.push(jsxElemToHTML(child, indent + 2));
      }

      return [startTag, ...children, endTag].join("\n");
    }
  }
};
