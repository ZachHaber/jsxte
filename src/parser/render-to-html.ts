import { jsxElemToHTML } from "./parse-template-to-html";

export const renderToHTML = <P>(
  component: (props: P) => JSX.Element,
  props: P
) => {
  return jsxElemToHTML(component(props));
};
