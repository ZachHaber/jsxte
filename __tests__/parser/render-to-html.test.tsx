import "../../src/jsx/jsx.types";
import { renderToHTML } from "../../src/parser/render-to-html";
// @ts-ignore
import { jsx, Fragment } from "../../src/jsx/jsx-runtime";
import { TJSXComponent } from "../../src/jsx/jsx.types";

describe("renderToHTML", () => {
  it("should correctly generate html from simple jsx", () => {
    const Component: TJSXComponent<{ title: string }> = (props) => {
      return (
        <div id="container" class={"bordered active"}>
          <h1>Hello World</h1>
          <h2>{props.title}</h2>
        </div>
      );
    };

    const html = renderToHTML(Component, { title: "Prop Title" });

    expect(html).toMatchSnapshot();
  });

  it("should correctly generate html from component base jsx structure", () => {
    const Header: TJSXComponent<{ title: string }> = (props) => {
      return <h2>{props.title}</h2>;
    };

    const Button: TJSXComponent<{ label: string; styles?: string }> = (
      props
    ) => {
      return <button style={props.styles}>{props.label}</button>;
    };

    const Template: TJSXComponent = ({ children }) => {
      return (
        <html>
          <head>
            <meta charset="utf-8" />
            <meta http-equiv="X-UA-Compatible" content="IE=edge" />
            <title>Page Title</title>
            <meta
              name="viewport"
              content="width=device-width, initial-scale=1"
            />
            <link
              rel="stylesheet"
              type="text/css"
              media="screen"
              href="main.css"
            />
          </head>
          <body>{children}</body>
        </html>
      );
    };

    const Content: TJSXComponent = () => {
      return (
        <>
          <Header title="Hello World!" />
          <input value="write here" />
          <Button label="Submit" />
        </>
      );
    };

    const App: TJSXComponent = () => {
      return (
        <Template>
          <div class="main-container">
            <Content />
          </div>
        </Template>
      );
    };

    const html = renderToHTML(App, {});

    expect(html).toMatchSnapshot();
  });
});
