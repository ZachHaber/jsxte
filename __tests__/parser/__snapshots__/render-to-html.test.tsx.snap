// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderToHTML should correctly generate html from component base jsx structure 1`] = `
"<html>
  <head>
    <meta charset=\\"utf-8\\">
    </meta>
    <meta http-equiv=\\"X-UA-Compatible\\">
    </meta>
    <title>
      Page Title
    </title>
    <meta name=\\"viewport\\">
    </meta>
    <link rel=\\"stylesheet\\">
    </link>
  </head>
  <body>
    <div class=\\"main-container\\">
        <h2>
          Hello World!
        </h2>
        <input value=\\"write here\\">
        </input>
        <button style>
          Submit
        </button>
    </div>
  </body>
</html>"
`;

exports[`renderToHTML should correctly generate html from simple jsx 1`] = `
"<div id=\\"container\\">
  <h1>
    Hello World
  </h1>
  <h2>
    Prop Title
  </h2>
</div>"
`;
